### PR TITLE
Sanitize thumbnail for Playlist contents

### PR DESCRIPTION
**Pull request by sangwoo108**
We didn't sanitize thumbnails as we use untrusted frame. But we need to sanitize the image as

* we are planing to change the top-most frame to normal chrome://frame
  * This might be needed as per a spec for showing playlist web ui on tabs

* We have a UI which shows thumbnails of detected media, which is not yet stored.

So we need facility to sanitize thumbnail from arbitrary sources.

Resolves https://github.com/brave/brave-browser/issues/24620

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/api_request_helper/api_request_helper.h [hunk 1]
```diff
@@ -197,20 +197,6 @@ class APIRequestHelper {
       const base::flat_map<std::string, std::string>& headers = {},
       const APIRequestOptions& request_options = {});
 
-  using DownloadCallback = base::OnceCallback<void(
-      base::FilePath,
-      const base::flat_map<std::string, std::string>& /*response_headers*/)>;
-
-  // TODO(petemill): Move Download (and OnDownload) to a separate module,
-  // it does not share much from this file, which is meant to be for JSON APIs.
```
Comment (sangwoo108) [inline components/api_request_helper/api_request_helper.h:205]: @petemill @nullhook , I removed download part as PlaylistThumbnailDownloader is the only caller.

File: browser/playlist/playlist_service_factory.cc [hunk 2]
```diff
@@ -53,26 +64,84 @@ class PlaylistServiceDelegateImpl : public PlaylistService::Delegate {
...
+        base::BindOnce(&PlaylistServiceDelegateImpl::EncodeAsPNG,
+                       weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+  }
+
  private:
+  scoped_refptr<base::SequencedTaskRunner> GetOrCreateTaskRunner() {
+    if (!task_runner_) {
+      task_runner_ = base::ThreadPool::CreateSequencedTaskRunner(
+          {base::MayBlock(), base::TaskPriority::USER_VISIBLE,
+           base::TaskShutdownBehavior::SKIP_ON_SHUTDOWN});
```
Comment (simonhong) [inline browser/playlist/playlist_service_factory.cc:102]: Q) `SKIP_ON_SHUTDOWN` is more proper over ``CONTINUE_ON_SHUTDOWN` here?
According to the [docs](https://source.chromium.org/chromium/chromium/src/+/main:base/task/task_traits.h;l=92;bpv=1;bpt=1), it'll blocks shutdown till it's completed.

**Review by simonhong — APPROVED** [review]
++ 👍🏼

**Review by mkarolin — APPROVED** [review]
`DEPS` ++

File: components/api_request_helper/api_request_helper.cc [hunk 3]
```diff
@@ -160,27 +160,6 @@ void APIRequestHelper::DeleteAndSendResult(Ticket iter,
...
-      request_options.enable_cache, false /*allow_http_error_result*/, headers);
-  iter->get()->url_loader_->DownloadToFile(
-      url_loader_factory_.get(),
-      base::BindOnce(&APIRequestHelper::OnDownload,
-                     weak_ptr_factory_.GetWeakPtr(), iter, std::move(callback)),
-      path);
-
-  return iter;
-}
-
```
Comment (petemill) [inline components/api_request_helper/api_request_helper.cc:183]: 👍

File: components/playlist/browser/playlist_thumbnail_downloader.cc [hunk 4]
```diff
@@ -59,84 +58,182 @@ void PlaylistThumbnailDownloader::DownloadThumbnail(
...
+  VLOG(2) << __func__ << " " << id << " : " << thumbnail_url.spec();
+  CancelDownloadRequest(id);
+
+  if (pause_download_for_testing_) {
+    url_loader_map_[id] = {};
+    return;
+  }
+
+  auto iter = CreateURLLoaderHandler(thumbnail_url);
+  iter->get()->DownloadToStringOfUnboundedSizeUntilCrashAndDie(
```
Comment (petemill) [inline components/playlist/browser/playlist_thumbnail_downloader.cc:87]: We don't want to set a download size limit?

**Comment by petemill** [comment]
Could you explain (or perhaps link to an issue) about the move from chrome-untrusted to chrome WebUI please?

File: components/playlist/browser/playlist_thumbnail_downloader.cc (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/browser/playlist_thumbnail_downloader.cc:87]: For now, I'm not sure how big it should be ☺️

**Comment by sangwoo108** [comment]
> Could you explain (or perhaps link to an issue) about the move from chrome-untrusted to chrome WebUI please?

Currently, Playlist web ui consists of 2 chrome-untrusted://frame, like figure below

image

We thought the top level frame(chrome-untrusted://playlist) should be also `untrusted` cause we might want to load thumbnails directly over the network, but we decided not to do that. We only shows thumbnail cached locally. So we can either make it back the plain `chrome://` or leave it as is.

When leaving it, we should check if the `chrome-untrusted://` url is revealed to users somewhere, as the url is not supposed to be exposed to users. For now I found one place from global media control

image

Repeating myself, we can either filter the URL or revert the frame to play `chrome` frame. But still we'll have the player area wrapped with the `chrome-untrusted://playlist-player`

**Comment by petemill** [comment]
Thanks for the explanation @sangwoo108. Now I understand fully. Seems good to move to trusted WebUI for future full-page ability, and to not have chrom-untrusted:// exposed. However, we might want to do both since having chrome://playlist display in the global media control seems weird as that is not a tab that the user is on (or will know the Url of), but one open i the sidebar. I wonder if we should investigate the effort to have the source of this item in the global media control say something like "Playlist Side Panel" and focus the panel that is playing it when clicking on the item...?

File: components/playlist/browser/playlist_thumbnail_downloader.cc (hunk 4 above)
Comment (petemill) [reply components/playlist/browser/playlist_thumbnail_downloader.cc:87]: Is there a disadvantage of setting something very high, like 5mb, just so a malicious endpoint can't crash the browser if they send 30gb of data?

**Review by petemill — APPROVED** [review]
Thanks for the explanations and good implementation. Seems good to go once you decide about what's appropriate for download size limit.

File: components/playlist/browser/playlist_thumbnail_downloader.cc (hunk 4 above)
Comment (sangwoo108) [reply components/playlist/browser/playlist_thumbnail_downloader.cc:87]: Yeah, makes sense. I googled what size is generally used for Youtube thumbnail and `og:tag image`. They say "1200~1280 * 720~1280". So 5mb would be good I think

**Comment by sangwoo108** [comment]
> I wonder if we should investigate the effort to have the source of this item in the global media control say something like "Playlist Side Panel"

Yes, I agree. AFAIK, global media control collects every media source states from web contents whatever the contents belongs to. Thus, showing "Playlist Side Panel" would be better.

**Review by deeppandya — APPROVED** [review]
LGTM

**Pull request merged**